text/0000-custom-panic-handlers.md

```diff
@@ -0,0 +1,111 @@
+
+Another concern is that some third-party library functions you have no control upon may panic on exceptional situations, and it is often not desirable to output debugging information to the user in this case.
+
+Currently, it is partially possible to customize the panic behavior by registering new callbacks with [`rt::unwind::register`](http://doc.rust-lang.org/std/rt/unwind/fn.register.html), however this solution is limited because it allows a maximum of 16 callbacks, and those are global, whereas the proposed solution uses a thread-local handler.
+
+# Detailed design
+
+Sample implementation: https://github.com/filsmick/rust/commit/8a5ae75e41863648f7c8cbbae3145e30bd260372
+
+In order to allow later extension of the data passed, as per [@sfackler's comment](https://github.com/rust-lang/rfcs/pull/1100#discussion_r33882931), handlers are functions accepting a `PanicData` parameter.
+
+
+The unstable [`Callback`](https://doc.rust-lang.org/std/rt/unwind/type.Callback.html) type is renamed to `PanicHandler` and changes from:
+``` rust
+fn(msg: &Any + Send, file: &'static str, line: u32)
+```
+
+to:
+```rust
+fn(panic_data: &PanicData)
+```
+where `PanicData` is an opaque struct with `msg`, `file` and `line` accessors, returning `&(Any + Send)`, `&'static str` and `u32`, respectively. This lets us add more fields to `PanicData` later, like backtrace data, for example.
+
+
+Handlers are thread-local. Since function pointers are `Copy`, a thread's panic handler is stored in a `Cell`.  
+The panic handler of a thread is changed by a setter function, `set_panic_handler`, which sets the inner value of the Cell to the new pointer.
+
+No handlers other than the default one would be added to `std`, because it is trivial to define more advanced handlers tailored to the needs of the program in user code. Common, reusable handlers can grow on crates.io without being tied to the standard library.
+
+`std::rt::unwind::begin_unwind_inner` would call the thread's panic handler before processing callbacks, without needing synchronization.
+
+If the user wishes not to log panics, he can define an empty function following the `Callback` signature and pass it to `set_panic_handler`.
+
+### Summary of the proposed changes
+
+* Add a thread-local handler to `rt::unwind`, defaulting to the [current default panic handler](https://github.com/rust-lang/rust/blob/2b8c9b12f91c0bf2c1e6278a5f803c2df3698432/src/libstd/panicking.rs#L28):
+```rust
+thread_local! { static ON_PANIC: Cell<Callback> = Cell::new(panicking::on_panic) }
+```
+
+* Add a function to `std::rt::unwind` or `std::thread` (see unresolved questions) to set a new panic handler:
+``` rust
+pub fn set_panic_handler(new_handler: PanicHandler) {
```

Comment:
This should probably be generalized to taking `F: PanicHandler` instead of forcing use of a function pointer, that will allow capturing data as well as using arbitrary closures to define handlers for panics.